locator/token_metadata.cc

```diff
@@ -1399,6 +1423,31 @@ future<> token_metadata_impl::calculate_pending_ranges_for_leaving(
     });
 }
 
+future<> token_metadata_impl::calculate_pending_ranges_for_replacing(
+        token_metadata& unpimplified_this,
+        abstract_replication_strategy& strategy,
+        lw_shared_ptr<std::unordered_multimap<range<token>, inet_address>> new_pending_ranges) {
+    if (_replacing_endpoints.empty()) {
+        return make_ready_future<>();
+    }
+    return do_with(_replacing_endpoints, strategy.get_address_ranges(unpimplified_this),
+            [this, new_pending_ranges] (std::unordered_map<inet_address, inet_address>& replacing_endpoints,
+                    std::unordered_multimap<inet_address, dht::token_range>& address_ranges) {
+        return do_for_each(replacing_endpoints, [this, new_pending_ranges, &address_ranges]
+                (const std::pair<gms::inet_address, gms::inet_address>& node) {
+            auto existing_node = node.first;
+            auto replacing_node = node.second;
+            return do_for_each(address_ranges, [this, new_pending_ranges, existing_node, replacing_node]
+                    (const std::pair<gms::inet_address, dht::token_range>& x) {
+                if (x.first == existing_node) {
+                    tlogger.debug("Node {} replaces {} for range {}", replacing_node, existing_node, x.second);
+                    new_pending_ranges->emplace(x.second, replacing_node);
+                }
+            });
+        });
+    });
+}
```

Comment:
I do not see that this function use futures internally, so first it can return `void` and second it can be simplified using `for()` loops. `do_with()` is also not needed.